src/Microsoft.AspNetCore.Razor.Evolution/DefaultRazorIRLoweringPhase.cs

```diff
@@ -120,14 +120,32 @@ public override void VisitEndTemplateBlock(TemplateBlockChunkGenerator chunkGene
             public override void VisitStartExpressionBlock(ExpressionChunkGenerator chunkGenerator, Block block)
             {
                 var value = new ContainerRazorIRNode();
-                Builder.Add(new CSharpExpressionIRNode()
+                var expressionNode = new CSharpExpressionIRNode()
                 {
                     Content = value,
-                    SourceLocation = block.Start,
-                });
+                };
+                Builder.Add(expressionNode);
 
                 var valueBuilder = RazorIRBuilder.Create(value);
                 _builders.Push(valueBuilder);
+
+                if (value.Children.Count > 0)
+                {
+                    var startLocation = value.Children[0].DocumentLocation;
+                    var contentLength = 0;
+
+                    for (var i = 0; i < value.Children.Count; i++)
+                    {
+                        contentLength += value.Children[i].DocumentLocation.ContentLength;
+                    }
+
+                    expressionNode.DocumentLocation = new MappingLocation(
```

Comment:
We were generating document locations incorrectly because the expression was saying its start was at the "@" symbol when that didn't exist as part of the IR.